Update Salt states to deploy the operator

slaperche-scality opened the pull request:
**Component**:

salt

**Context**: 

In order to create PV (PersistentVolume), we need to have an operator up and running.

**Summary**:

We use the YAML files generated by the Operator SDK as Salt states, thanks to our custom renderer.

**Acceptance criteria**: 

After bootstrap, check that
- the Volume CRD exists
- the operator is running
- the StorageClass exists

---

Closes: #1411

slaperche-scality committed:
salt/deployment: deploy the storage operator

We import the YAML generated by the operator SDK, but takes care of
adding tolerations and the right image name for the Deployment manifest.
We also add a namespace for the ServiceAccount.

Refs: #1411
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality committed:
salt/deployment: deploy Volume CRD

Refs: #1411
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality committed:
buildchain/salt: deploy the new SLS files

Refs: #1411
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

bert-e commented:
# Hello slaperche-scality,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/metalk8s/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @teddyandrieux

* @monpote

* @jbertran

* @sayf-eddine-scality

* @nicolast

* @cmonfort

* @slaperche-scality

* @thomasdanan

* @zempashi

* @lucieleonard

* @gdemonet

* @carlito-scality

* @nootal

* @alexandre-allard-scality

slaperche-scality commented on salt/metalk8s/addons/volumes/storage-classes.sls:
```diff
@@ -0,0 +1,14 @@
+apiVersion: storage.k8s.io/v1
+kind: StorageClass
+metadata:
+  name: metalk8s-prometheus
```
not sure if it's a good name

NicolasT commented on salt/metalk8s/addons/volumes/volume-crd.sls:
```diff
@@ -0,0 +1,77 @@
+#!jinja | kubernetes kubeconfig=/etc/kubernetes/admin.conf&context=kubernetes-admin@kubernetes
```
Hmh, could this be generated from the file as generated by `operator-sdk generate ...` (i.e., the file in `storage-operator/disk/...`) at build time, instead of copying it over?

slaperche-scality replied:
Yup I plan to do it in a later commit (PR still in draft ^^), first I try to make it work like that, then I'll make sure we reuse the one generated by the SDK to avoid desync issue.

slaperche-scality replied:
Done

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @teddyandrieux

* @monpote

* @jbertran

* @sayf-eddine-scality

* @nicolast

* @cmonfort

* @slaperche-scality

* @thomasdanan

* @zempashi

* @lucieleonard

* @gdemonet

* @carlito-scality

* @nootal

* @alexandre-allard-scality





The following reviewers are expecting changes from the author, or must review again:

* @nicolast

slaperche-scality committed:
buildchain/image: add operator image into version.py

We now use a central place to manage the versions (version.py) and all
the images should be declared there (because it's used to generate a
file for Salt).

It seems that the `storage-operator` wasn't updated to use this file,
now it's done.

Refs: #1411
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality committed:
salt/deployment: deploy the StorageClass for Prometheus

Refs: #1411
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality committed:
salt/renderer: add support for StorageClass

Refs: #1411
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality committed:
storage-operator/deploy: backport changes to the original files

In previous commits the content of those files was modified and copied
into the Salt state directly.

Now the we know it works, it's time to update the original files.
Next step (commit) is to use those original files to generate the SLS
ones.

Refs: #1411
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality committed:
buildchain/salt: generate SLS from SDK files

This is especially important for the Volume CRD that will certainly
evoles: we must keep the version used by Salt up-to-date.
The best way to do so is to use the generated file to generate the Salt
SLS.

Refs: #1411
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality committed:
test/volume: add tests for the operator deployment

For now we simply test that we have a storage class and that the
operator's pod is up.

Closes: #1411
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

TeddyAndrieux commented on salt/metalk8s/addons/volumes/storage-operator.sls.in:
```diff
@@ -0,0 +1,10 @@
+---
+@@RoleBinding
+---
+@@Deployment
```
Do we have a reason to put yaml in directory `storage-operator/deploy` instead of directly here ?

slaperche-scality replied:
When you ugrade operator-sdk, there is no way to upgrade your generated project.

You need to re-create one from scratch with the new version, do a diff with the new scaffolding and your current code and then apply the changes manually.

Keeping files where the SDK generated them will ease the painful manual process.

NicolasT replied:
Also, some of those files are regenerated by `operator-sdk generate` into the `deploy/` folder of your operator, and we don't want to need to manually keep the copy in sync.

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @teddyandrieux

* @monpote

* @jbertran

* @sayf-eddine-scality

* @nicolast

* @cmonfort

* @slaperche-scality

* @thomasdanan

* @zempashi

* @lucieleonard

* @gdemonet

* @carlito-scality

* @nootal

* @alexandre-allard-scality

TeddyAndrieux commented on salt/_states/metalk8s_kubernetes.py:
```diff
@@ -1887,3 +1887,33 @@ def apiservice_absent(name, **kwargs):
+        ret['result'] = True
+        ret['changes'][name] = {'old': {}, 'new': res}
+    else:
+        ret['result'] = True if not __opts__['test'] else None
```
Minor: not agree with this one. 
Global salt guideline

Test mode enabled:
- Failure -> result = False
- Nothing to do -> result = True
- Something will be changed -> result = None

So here you should always return True as all good

slaperche-scality replied:
Ok, will do. Thanks for the explanation.

(We probably want to fix `namespace_present` and `serviceaccount_present` but that's probably target an older branch so won't be done here).

TeddyAndrieux commented on salt/_states/metalk8s_kubernetes.py:
```diff
@@ -1887,3 +1887,33 @@ def apiservice_absent(name, **kwargs):
 
     return ret
+
+def storageclass_present(
```
`E302: expected 2 blank lines, found 1`
:stuck_out_tongue:

slaperche-scality replied:
Meh. Done

TeddyAndrieux commented on salt/_states/metalk8s_kubernetes.py:
```diff
@@ -1887,3 +1887,34 @@ def apiservice_absent(name, **kwargs):
+        ),
+        ret['result'] = True
+        ret['changes'][name] = {'old': {}, 'new': res}
+    else:
```
Also here we do not handle the case where storageclass exists but do not have the right "stuff".
I mean if we change the `mount_options` for example and run the state a second time we will not update it.
This is what we want ?

NicolasT replied:
Yes, we only use the `StorageClass` info when we provision the volume. If for whatever reason someone changes it later (say, change the `fsType`), there's really not much we can do!

slaperche-scality replied:
Yeah, changing the value of an SC isn't going to impact the existing volumes.

But if:
- in 2.4, I have a StorageClass `Foo` that defines `rw` as `mountOptions`.
- in 2.5, I have updated the StorageClass `Foo` to have `rw,some_new_stuff` as `mountOptions`.

Of course I don't want existing volume to be impacted, but we probably want new volumes to be mounted with `some_new_stuff`.
In that case, I should update the StorageClass if it exists no?

Otherwise `SC` are never "upgraded".

slaperche-scality replied:
Done

TeddyAndrieux reviewed (approved):
Salt code LGTM but I not tested it, Maybe you want to wait for someone else to test it, as you want

slaperche-scality commented:
> Salt code LGTM but I not tested it, Maybe you want to wait for someone else to test it, as you want

I've added an automated test that checks that after a bootstrap you have the StorageClass created and the Operator running, so at least it's not utterly broken ^^

Thanks for the review!

slaperche-scality commented:
/approve

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/2.4`



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/2.4`



The following branches have **NOT** changed:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`



Please check the status of the associated issue None.

Goodbye slaperche-scality.
